cryptoki/src/mechanism/kbkdf.rs

```diff
@@ -0,0 +1,371 @@
+///
+/// * [`PrfDataParamType::IterationVariable`] is required for the KDF in all modes.
+///   * In counter-mode, [`PrfDataParamType::IterationVariable`] must contain [`KbkdfCounterFormat`].
+///     In feedback- and double pipeline-mode, it must contain [`None`].
+/// * [`PrfDataParamType::Counter`] must not be present in counter-mode, and can be present at most
+///   once in feedback- and double-pipeline modes.
+/// * [`PrfDataParamType::DkmLength`] can be present at most once, in any mode.
+/// * [`PrfDataParamType::ByteArray`] can be present any amount of times, in any mode.
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct PrfDataParam<'a> {
+    inner: CK_PRF_DATA_PARAM,
+    /// Marker type to ensure we don't outlive the data
+    _marker: PhantomData<&'a [u8]>,
+}
+
+impl<'a> PrfDataParam<'a> {
+    /// Construct data parameter for input of the PRF internal to the KBKDF.
+    ///
+    /// # Arguments
+    ///
+    /// * `type_` - The specific type and parameters for the data parameter.
+    pub fn new(type_: PrfDataParamType<'a>) -> Self {
+        Self {
+            inner: match type_ {
+                PrfDataParamType::IterationVariable(None) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_ITERATION_VARIABLE,
+                    pValue: ptr::null_mut(),
+                    ulValueLen: 0,
+                },
+                PrfDataParamType::IterationVariable(Some(counter_format)) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_ITERATION_VARIABLE,
+                    pValue: counter_format as *const _ as *mut _,
+                    ulValueLen: size_of::<CK_SP800_108_COUNTER_FORMAT>() as CK_ULONG,
+                },
+                PrfDataParamType::Counter(counter_format) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_COUNTER,
+                    pValue: counter_format as *const _ as *mut _,
+                    ulValueLen: size_of::<CK_SP800_108_COUNTER_FORMAT>() as CK_ULONG,
+                },
+                PrfDataParamType::DkmLength(dkm_length_format) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_DKM_LENGTH,
+                    pValue: dkm_length_format as *const _ as *mut _,
+                    ulValueLen: size_of::<CK_SP800_108_DKM_LENGTH_FORMAT>() as CK_ULONG,
+                },
+                PrfDataParamType::ByteArray(data) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_BYTE_ARRAY,
+                    pValue: data.as_ptr() as *mut _,
+                    ulValueLen: data
+                        .len()
+                        .try_into()
+                        .expect("length of PRF data parameter does not fit in CK_ULONG"),
+                },
+            },
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Container for information on an additional key to be derived.
+#[derive(Debug)]
+pub struct DerivedKey {
+    template: Pin<Box<[CK_ATTRIBUTE]>>,
```

Comment:
Just out of curiosity why this is pinned?

Reply:
Since this is a private field, I think it's fine with only the `Box` but writing the reasoning in a comment above it :)